### ISSUE TITLE
md-backdrop what positioning should it have?

### ISSUE THREAD

**Issue by fant0m** [angular/material#3774]
Hello,
I just don't know, 
**backdrop:**
+  * use absolute positioning and anchors ([87bdf36a](https://github.com/angular/material/commit/87bdf36a))
+  * set css `position: fixed` to `md-backdrop` ([4011df9c](https://github.com/angular/material/commit/4011df9c))

so should it have fixed position or absolute? In my opinion it should have fixed. I'm asking if you made a mistake or if it gonna stay with absolute position.

**Comment by ThomasBurleson** [comment]
@fant0m - This is not a mistake. Several issues manifested with backdrop fixed. With absolute **and** anchors top, bottom, right, left... makes the backdrop always fill the viewport.

Can you provide details on how `fixed` solves issues that the current settings do not ?

**Comment by fant0m** [comment]
Hm, I made a demo http://codepen.io/anon/pen/GJBrXP?editors=101, scroll down and click on button

**Comment by bwegrzyn** [comment]
I'm experiencing this issue as well. As soon as the page scrolls, the backdrop no longer fills the viewport.

**Comment by metamatt** [comment]
This is breaking menu dismissing, as far as I can tell. As soon as I scroll the page, the md-backdrop scrolls with it. It has "position: absolute; top left bottom right: 0". The body element gets some funny styles applied when a menu is open, which I think is the work of this code:

```
            // Converts the body to a position fixed block and translate it to the proper scroll
            // position
            function disableBodyScroll() {
              var htmlNode = body.parentNode;
              var restoreHtmlStyle = htmlNode.getAttribute('style') || '';
              var restoreBodyStyle = body.getAttribute('style') || '';
              var scrollOffset = $mdUtil.scrollTop(body);
              var clientWidth = body.clientWidth;

              if (body.scrollHeight > body.clientHeight) {
                applyStyles(body, {
                  position: 'fixed',
                  width: '100%',
                  top: -scrollOffset + 'px'
                });

                applyStyles(htmlNode, {
                  overflowY: 'scroll'
                });
              }
```

This leads to a body element that looks like `<body style="position: fixed; width: 100%; top: -704px;">` after scrolling, and a child `<md-backdrop>` element that's scrolled off the top of the screen.

As a side effect of this, md-menus do not auto-dismiss when you click outside them on the area of the screen that was newly revealed by the scrolling.

@ThomasBurleson could you please take another look? I think there's a genuine bug here, not just a question about "what should the position attribute be".

From what I can see the point of `disableBodyScroll` is indeed to disable scrolling while a menu is open -- if I remove the styles added by `disableBodyScroll` then the md-backdrop works but the whole body scrolls around which is disconcerting. There must be some way to get both behaviors. Maybe md-backdrop needs a top corresponding to the body scroll offset?

**Comment by metamatt** [comment]
Here's a quick hack patch which fixes the problem that I see in the md-menu case:

```
--- a/web/jspm_packages/github/angular/bower-material@0.10.1/angular-material.js
+++ b/web/jspm_packages/github/angular/bower-material@0.10.1/angular-material.js
@@ -9122,6 +9122,8 @@ function MenuProvider($$interimElementProvider) {

       if (opts.backdrop) {
         $mdTheming.inherit(opts.backdrop, opts.parent);
+        var scrollOffset = $('body').position().top;
+        opts.backdrop.css('bottom', scrollOffset);
         opts.parent.append(opts.backdrop);
       }
       showMenu();
```

This is just to demonstrate the problem, certainly not the right fix. It's a little tricky because presumably the same problem and fix would apply to all uses of `<md-backdrop>`, not just the menu-related uses, so we want to fix `$mdUtil.createBackdrop()`, but in the case of menus at least, the code order is:

* create backdrop (but don't insert into DOM)
* then disable parent scroll (which might call `disableBodyScroll()`, adding a nonzero "top" to the body element)
* then insert the backdrop element into the DOM

so `createBackdrop` can't see the change to the `<body>` element yet, and `disableBodyScroll` can't see the `<md-backdrop>` element, so they can't just reach in and know about each other -- which might be a bad idea anyway -- something else is going to have to coordinate these.

**Comment by robertmesserle** [comment]
I am not seeing any issue in the CodePen demo.  Is this still an issue?

**Comment by robertmesserle** [comment]
Closing for now - please reopen if this is still an issue and provide a codepen or instructions to recreate.

**Comment by metamatt** [comment]
@robertmesserle I don't have permission to reopen but I believe this is still an issue, but codepen is doing something to mask it such that you can't repro at codepen.

If I try using AM in a real app, the `<html>` and `<body>` elements have `position: relative, height: 100%` and are pinned to the size of the browser window but if I scroll the browser window they scroll offscreen. If I try the codepen, the `<html>` and `<body>` elements (for the penned app, inside the iframe) get positioned completely differently -- according to chrome devtools, the `html, body { position: relative; height: 100%;` styles are grayed out, and the bounding box corresponds roughly to the editor part of the codepen UI, not the client presentation rectangle. And, crucially, does not scroll when I scroll the document.

Disclaimer: I'm no CSS expert. Maybe the styles I'm pointing at aren't the root cause. But I definitely see a bug easily reproducible in my real app that does what that codepen does, and in both my real app and in a plunkr.co demo the root body/html elements get that funny "scrolls offscreen" behavior, and in codepen they do not, so I think: try repro'ing without codepen.

**Comment by robertmesserle** [comment]
@metamatt If you are seeing this in a Plunker demo, that is fine as long as I have a way of seeing this issue occurring.

**Comment by metamatt** [comment]
Thanks. I just tried moving the dialog demo from codepen to plunker and still didn't see the behavior that was bugging me in the real app. I'll try to isolate it to a working repro in either codepen or plunker and get back to you.

**Comment by robertmesserle** [comment]
@metamatt Also, if your app is available for me to view, I could try to debug it there as well.

**Comment by metamatt** [comment]
OK, I couldn't repro in plunker so I dug into the code to see what was different between my app case and the plunker case even using the same version (still on 0.10.1). I think this got fixed *inside* the 0.10.1 release. Sorry to waste your time but if you have a minute maybe you could explain or link me to something about your release management practices that would help me and others reliably obtain the right code?

Here's what I'm seeing: in the correct case, when a menu is onscreen there's a `md-backdrop.md-menu-backdrop` with style `position: fixed !important`. In my broken case, it's lacking that style. I see that style in [backdrop.scss on master](https://github.com/angular/material/blame/master/src/components/backdrop/backdrop.scss#L16), dating to d9ba0e1 which is in 0.11 but not 0.10.1 according to github's lineage tracer.

However, if I look at the bower-material blobs generated from that file, I see that `position: fixed !important` style showed up sometime *during* the 0.10.1 release cycle, at least judging from https://github.com/angular/bower-material/blob/44f067ad06830daf69a173706e7dee92edb4b1fa/modules/closure/backdrop/backdrop.css (first release tagged 0.10.1 in the bower-material repo) and https://github.com/angular/bower-material/blob/6cd04759a57db62fae6eb99a5a2dc85c7b427749/modules/closure/backdrop/backdrop.css (last release tagged 0.10.1 in the bower-material repo).

For my live project, I got the source via jspm on August 11, and haven't updated it since. The backdrop.css in my tree (as well as the concatenated version in angular-material.css) do not have that style.

This still doesn't explain why if I try a minimal repro case using the 44f067ad06830daf69a173706e7dee92edb4b1fa version of the source (via https://rawgit.com/angular/bower-material/44f067ad06830daf69a173706e7dee92edb4b1fa/angular-material.js) it has the right style and works correctly. But it probably explains why most other people aren't able to repro this.

Anyway, I'm willing to bet that the bug has long been fixed and I just have an old snapshot of the code. I'm happy to see this get closed and when I get around to updating to 0.11 or 0.12 I'll file a new bug if I see any related issue return with my real codebase.

Is there any protocol I should be following to know when to grab a new snapshot of the bower-material artifacts inside the same release version? I was surprised to see so many different 0.10.1 tags with different actual content.

**Comment by robertmesserle** [comment]
The tagging may be a bit misleading.  Once we release 0.10.1, the official version of the code changes to 0.10.1, but each commit after that is not part of that commit.  So a tag in bower that looks like `v0.10.1-{{SHA}}` is **NOT** included in `0.10.1` and will be rolled into the next release.
